### ISSUE TITLE
CodeQL analysis broken after introducing IBM MQ components

### ISSUE THREAD

**Issue by antoineco** [triggermesh/triggermesh#410]
https://github.com/triggermesh/triggermesh/runs/4750579570?check_suite_focus=true

>  ```
> go build ./cmd/...
>
> [...]
>
> # github.com/ibm-messaging/mq-golang/v5/ibmmq
> Error: ../../../go/pkg/mod/github.com/ibm-messaging/mq-golang/v5@v5.2.4/ibmmq/mqi.go:54:10: fatal error: cmqc.h: No such file or directory
>   54 | #include <cmqc.h>
>      |          ^~~~~~~~
> compilation terminated.
> ```

Maybe we need to install extra libraries as part of the workflow.

cc. @tzununbekov

**Comment by antoineco** [comment]
I also see this warning in `golangci-lint`'s logs:

```
  level=warning msg="[runner] Can't run linter goanalysis_metalinter: buildir: failed to load package ibmmq: could not load export data: no export data for \"github.com/ibm-messaging/mq-golang/v5/ibmmq\""
```

**Comment by tzununbekov** [comment]
I need a piece of advice here.

The workflow is failing because we're trying to go-build everything in `cmd/`:
https://github.com/triggermesh/triggermesh/blob/5f13a11a6ed08cd4de63b64f9c1f44910e6ea018/.github/workflows/codeql.yaml#L49
but as we already established - IBM MQ cannot be built in this manner: https://github.com/ibm-messaging/mq-golang#using-the-package
Right now I see a couple of possible solutions that look equally ugly for me:
- Install IBM MQ Redistributable client on CI machine to make `go build` work there. The steps to do that can be borrowed from [here](https://github.com/triggermesh/triggermesh/blob/93cf325531f27195635b62a63d3ddc9cbb24c706/cmd/ibmmqsource-adapter/Dockerfile#L5-L14). Cons - scales badly for any other custom-built adapter that we may add in the future.
- Move custom-built adapters to their own root-level directory thereby excluding them from `go build ./cmd/...` command. Cons - excluding components from the common linting and analysis workflows. 
- Keep all adapters in `cmd/` but add some sort of components exclusions list in failing workflows. The downside is the same as for the previous one - not having the lining and analysis reports.

@antoineco what would you recommend here?

**Comment by antoineco** [comment]
@tzununbekov this workflow is a bit special, it's a security oriented flow that runs once a week. We do want to find security issues in _all_ the code, but it doesn't matter if the CI looks ugly, as long as CodeQL can trace the build and do its magic.

My preference would be:
- include the required libs in the CodeQL workflow to prevent the build from failing (note: if we end up with a lot of components in this situation we can analyse them separately in the future)
- Ignore MQ integrations in the linting workflow.

**Comment by antoineco** [comment]
Actually the linter seems to be broken by this change too.
There are obvious mistakes that were not reported in another PR.

**Comment by tzununbekov** [comment]
I'm working on it right now!

**Comment by antoineco** [comment]
golangci-lint's `skip-dirs` option didn't work for me unfortunately.

**Comment by tzununbekov** [comment]
I don't know if it is specific for my system only, but I had to set `CGO_ENABLED` to 0 and add skip-files/skip-dirs to make linter happy:
```
CGO_ENABLED=0 golangci-lint run --skip-files pkg/targets/adapter/ibmmqtarget/mq/mqutil.go,pkg/sources/adapter/ibmmqsource/mq/mqutil.go --skip-dirs pkg/targets/adapter/confluenttarget
```

**Comment by antoineco** [comment]
Interesting.

I see you excluded `confluenttarget`, but that one didn't cause any failure for us. Only MQ and XSLT seem to.

**Comment by tzununbekov** [comment]
Yeah, I always had some issues with confluent build/linting, must be something with my machine. Had to add it to `skip-dirs` for local tests, it won't be there when I'll find a working CI config.

**Comment by antoineco** [comment]
I think our only choice is to generate a list of packages using `go list`, exclude specific (problematic) packages, and pass it to our tools.

As long as we keep using `./...`, the Go toolchain _will_ attempt to compile code.

> We can't not build skipped packages: it's the limitation of go/packages library.
>
> -- https://github.com/golangci/golangci-lint/issues/489#issuecomment-485235860

**Comment by tzununbekov** [comment]
@antoineco hey, I've added linter configs to the workflow and I got curious output:

https://github.com/tzununbekov/triggermesh/runs/4772220308?check_suite_focus=true

Now CI gets a lot of stylcheck errors in kafka package, just like I had locally. I'm wondering if it is possible that before we somehow implicitly skipped `confluenttarget` package?
And btw, looks like MQ issue is fixed there.

**Comment by antoineco** [comment]
Does it mean we'll need to pass these to `ko` too?

**Comment by tzununbekov** [comment]
I suppose no. We should add build tag for golangci runs ([here](https://github.com/tzununbekov/triggermesh/blob/ibmmq-build-fix/.golangci.yaml#L14-L15)) and then add this tag exclusions in the files that we want to skip ([example](https://github.com/tzununbekov/triggermesh/blob/ibmmq-build-fix/cmd/ibmmqsource-adapter/main.go#L1)). All other builds without this tag should not be affected.

**Comment by tzununbekov** [comment]
https://github.com/triggermesh/triggermesh/pull/417

**Comment by antoineco** [comment]
Closing because the CI workflows are now completing again:

- CodeQL: https://github.com/triggermesh/triggermesh/runs/4776260938?check_suite_focus=true
- Linter: https://github.com/triggermesh/triggermesh/runs/4776551403?check_suite_focus=true
